internal/pkg/api/handlePGPRequest.go

```diff
@@ -0,0 +1,177 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/fleet-server/v7/internal/pkg/cache"
+	"github.com/elastic/fleet-server/v7/internal/pkg/config"
+	"github.com/rs/zerolog"
+	"go.elastic.co/apm/v2"
+)
+
+const (
+	defaultKeyName        = "default.pgp"
+	defaultKeyPermissions = 0o0600
+)
+
+var (
+	ErrTLSRequired    = errors.New("api call requires a TLS connection")
+	ErrPGPPermissions = fmt.Errorf("pgp key permissions are not %#o", defaultKeyPermissions)
+	ErrUpstreamStatus = errors.New("upstream http server status error")
+)
+
+type PGPRetrieverT struct {
+	bulker bulk.Bulk
+	cache  cache.Cache
+	cfg    config.PGP
+}
+
+func NewPGPRetrieverT(cfg *config.Server, bulker bulk.Bulk, c cache.Cache) *PGPRetrieverT {
+	return &PGPRetrieverT{
+		bulker: bulker,
+		cache:  c,
+		cfg:    cfg.PGP,
+	}
+}
+
+func (pt *PGPRetrieverT) handlePGPKey(zlog zerolog.Logger, w http.ResponseWriter, r *http.Request, _, _, _ int) error {
+	if r.TLS == nil {
+		return ErrTLSRequired
+	}
+	key, err := authAPIKey(r, pt.bulker, pt.cache)
+	if err != nil {
+		return err
+	}
+	zlog = zlog.With().Str(LogEnrollAPIKeyID, key.ID).Logger()
+	ctx := zlog.WithContext(r.Context())
+
+	p, err := pt.getPGPKey(ctx, zlog)
+	if err != nil {
+		return err
+	}
+
+	_, err = w.Write(p)
+	return err
+}
+
+// getPGPKey will return the PGP key bytes
+//
+// First the local cache will be checked
+// If it's not found in the cache, we attempt to read from disk
+// If it's found we set the cache and return the bytes
+// If it's not found on disk we attempt to retrieve the upstream key
+// If that succeeds we set the cache then write to disk (with best effort).
+func (pt *PGPRetrieverT) getPGPKey(ctx context.Context, zlog zerolog.Logger) ([]byte, error) {
+	// key that will be retrieved, if this ever changes we should ensure that it can't be used as part of an injection attack as it is joined as part of the filepath for operations.
+	key := defaultKeyName
+
+	span, ctx := apm.StartSpan(ctx, "getPGPKey", "process")
+	span.Context.SetLabel("key", key)
+	defer span.End()
+
+	p, ok := pt.cache.GetPGPKey(key)
+	if ok {
+		return p, nil
+	}
+	p, err := pt.getPGPFromDir(ctx, key)
+
+	// successfully retrieved from disk
+	if err == nil {
+		pt.cache.SetPGPKey(key, p)
+		return p, nil
+	}
+
+	// file exists but the read failed
+	if !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+
+	// file does not exist
+	p, err = pt.getPGPFromUpstream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	pt.cache.SetPGPKey(key, p)
+	pt.writeKeyToDir(ctx, zlog, key, p)
```

Comment:
If we write the upstream key to disk, this mean if that key is updated user will have to manually update it on their fleet-server?

Reply:
Yes, if the upstream key is updated the user will need to manually update the key on disk (or remove it)